src/core/ext/transport/cronet/transport/cronet_transport.c

```diff
@@ -51,612 +51,918 @@
+    if (grpc_cronet_trace) gpr_log(__VA_ARGS__); \
+  }
+
+/* TODO (makdharma): Hook up into the wider tracing mechanism */
 int grpc_cronet_trace = 1;
 
-// Cronet transport object
+enum OP_RESULT {
+  ACTION_TAKEN_WITH_CALLBACK,
+  ACTION_TAKEN_NO_CALLBACK,
+  NO_ACTION_POSSIBLE
+};
+
+/* Used for printing debug */
+const char *op_result_string[] = {"ACTION_TAKEN_WITH_CALLBACK",
+                                  "ACTION_TAKEN_NO_CALLBACK",
+                                  "NO_ACTION_POSSIBLE"};
+
+enum OP_ID {
+  OP_SEND_INITIAL_METADATA = 0,
+  OP_SEND_MESSAGE,
+  OP_SEND_TRAILING_METADATA,
+  OP_RECV_MESSAGE,
+  OP_RECV_INITIAL_METADATA,
+  OP_RECV_TRAILING_METADATA,
+  OP_CANCEL_ERROR,
+  OP_ON_COMPLETE,
+  OP_FAILED,
+  OP_SUCCEEDED,
+  OP_CANCELED,
+  OP_RECV_MESSAGE_AND_ON_COMPLETE,
+  OP_READ_REQ_MADE,
+  OP_NUM_OPS
+};
+
+const char *op_id_string[] = {"OP_SEND_INITIAL_METADATA",
+                              "OP_SEND_MESSAGE",
+                              "OP_SEND_TRAILING_METADATA",
+                              "OP_RECV_MESSAGE",
+                              "OP_RECV_INITIAL_METADATA",
+                              "OP_RECV_TRAILING_METADATA",
+                              "OP_CANCEL_ERROR",
+                              "OP_ON_COMPLETE",
+                              "OP_FAILED",
+                              "OP_SUCCEEDED",
+                              "OP_CANCELED",
+                              "OP_RECV_MESSAGE_AND_ON_COMPLETE",
+                              "OP_READ_REQ_MADE",
+                              "OP_NUM_OPS"};
+
+/* Cronet callbacks */
+
+static void on_request_headers_sent(cronet_bidirectional_stream *);
+static void on_response_headers_received(
+    cronet_bidirectional_stream *,
+    const cronet_bidirectional_stream_header_array *, const char *);
+static void on_write_completed(cronet_bidirectional_stream *, const char *);
+static void on_read_completed(cronet_bidirectional_stream *, char *, int);
+static void on_response_trailers_received(
+    cronet_bidirectional_stream *,
+    const cronet_bidirectional_stream_header_array *);
+static void on_succeeded(cronet_bidirectional_stream *);
+static void on_failed(cronet_bidirectional_stream *, int);
+static void on_canceled(cronet_bidirectional_stream *);
+static cronet_bidirectional_stream_callback cronet_callbacks = {
+    on_request_headers_sent,
+    on_response_headers_received,
+    on_read_completed,
+    on_write_completed,
+    on_response_trailers_received,
+    on_succeeded,
+    on_failed,
+    on_canceled};
+
+/* Cronet transport object */
 struct grpc_cronet_transport {
   grpc_transport base; /* must be first element in this structure */
   cronet_engine *engine;
   char *host;
 };
-
 typedef struct grpc_cronet_transport grpc_cronet_transport;
 
-enum send_state {
-  CRONET_SEND_IDLE = 0,
-  CRONET_REQ_STARTED,
-  CRONET_SEND_HEADER,
-  CRONET_WRITE,
-  CRONET_WRITE_COMPLETED,
+struct read_state {
```

Comment:
it may be worth, especially on memory-limited mobiles, to reorder the struct fields in decreasing order of memory aligment. Quoting from http://www.catb.org/esr/structure-packing/#_structure_reordering: 

> The simplest way to eliminate slop is to reorder the structure members by decreasing alignment. That is: make all the pointer-aligned subfields come first, because on a 64-bit machine they will be 8 bytes. Then the 4-byte ints; then the 2-byte shorts; then the character fields.